### ISSUE TITLE
Feature Request: Override macOS appearance (light/dark) based on VSCode theme

### ISSUE THREAD

**Issue by rinsuki** [microsoft/vscode#120420]
I'm always using Dark+ Theme at VSCode (regardless of macOS light/dark appearance), but when `window.titleBarStyle` set to `native`, title bar style is mismatching with VSCode theme (if system-wide appearance is light). looks like this:

<img width="1045" alt="image" src="https://user-images.githubusercontent.com/6533808/113414206-67a4f080-93f7-11eb-9bf3-8492c5eeba12.png">

Electron has `require("electron").nativeTheme.themeSource` API and `BrowserWindow#vibrancy` API that can overrides system appearance per app/window, so I injected `require("electron").nativeTheme.themeSource = "dark"` to `Visual Studio Code.app/Contents/Resources/app/out/main.js` (yeah, its unsupported, i know) and looks pretty great.

<img width="1045" alt="image" src="https://user-images.githubusercontent.com/6533808/113414444-fc0f5300-93f7-11eb-80a2-a82399380ee3.png">

However, I think code injection is not supported, possibly corrupting the installed VS Code, and disabling it with every VS Code update.

so I'll suggest new settings that overrides macOS appearance based on VSCode theme.

**Note:** I think this new settings may conflicts with `window.autoDetectColorScheme` and `window.autoDetectHighContrast`.

**Comment by bpasero** [comment]
This is pretty cool, I was not aware we can change the title style to dark style.

@deepak1556 is this API robust to use? Do you see issues with it?

**Comment by bpasero** [comment]
Hm, one issue is that theme settings can change per workspace but from the main side we have no way of reading the workspace settings...

**Comment by bpasero** [comment]
I tested this and it seems to work really well on macOS at least. Still, I wonder if this needs to be a separate setting that the user can control it and wether we should not do this automatically.

@aeschli do you have thoughts on this? I know you have added theming related code and settings that automatically switch the theme in VSCode between dark and light depending on the OS setting. Here we would introduce a setting for the user to "hard-code" the appearance of the application to be either dark or light. I would think this setting can only be application scoped and not per workspace though because we need to resolve this from the main side.

**Comment by aeschli** [comment]
It looks good!

But it means we no longer get events when the OS change schemes, right?

BTW when the themeservice sets a new theme, the main process gets an event: https://github.com/microsoft/vscode/blob/main/src/vs/platform/theme/electron-main/themeMainService.ts#L31

**Comment by bpasero** [comment]
@aeschli 

> But it means we no longer get events when the OS change schemes, right?

I think the idea is that you must use `nativeTheme.shouldUseDarkColors` whenever you are making a decision about which theme is being used. Not sure we do that already. https://github.com/electron/electron/blob/11-x-y/docs/api/native-theme.md#nativethemethemesource explains the impact quite well.

> BTW when the themeservice sets a new theme, the main process gets an event

Yeah but that is probably too late for a smooth startup. Imagine a workspace configures the theme to be light but the default setting is dark, we would probably still set dark first only to update to light then.

**Comment by aeschli** [comment]
If the setting `window.autoDetectColorScheme` is true then [we use](https://github.com/microsoft/vscode/blob/main/src/vs/platform/native/electron-main/nativeHostMainService.ts#L63) `shouldUseDarkColors` and`nativeTheme.on('updated,..`, to get information about the OS scheme. 
But from reading the spec and what @rinsuki also mentioned, I believe once we set `nativeTheme.themeSource = "dark"`, `shouldUseDarkColors` will become and remain true and we no longer get to know what the OS scheme is and when it changes. Can you verify if that's like that? That would be a problem as we could no longer support `window.autoDetectColorScheme` .

> Yeah but that is probably too late for a smooth startup. Imagine a workspace configures the theme to be light but the default setting is dark, we would probably still set dark first only to update to light then.

It's the same as for the initial window background. The ThemeMainService listens to theme changes and remembers the background of the last used theme. Yes, that can result in flickering when themes are configured in the workspace settings.

**Comment by bpasero** [comment]
Hm, I wonder if we should actually deprecate `window.autoDetectColorScheme` and have a new `window.colorScheme` setting with these choices:
* `off`
* `system`: check OS theme
* `theme`: pick a color scheme that matches VSCode theme (dark or light)
* `light`: hardcode
* `dark`: hardcode

This would solve the issue of conflicting settings.

**Comment by aeschli** [comment]
Can you explain what that new setting does and how it replaces the old settings?

![image](https://user-images.githubusercontent.com/6461412/114188388-628cf780-9949-11eb-8feb-94d522a19281.png)

**Comment by bpasero** [comment]
Yeah you are right, both are probably too different to merge into one setting because "old" setting is for changing the theme in VSCode based on OS configuration and the "new" setting is for changing the OS configuration (for the window) based on the VSCode theme.

**Comment by aeschli** [comment]
Thinking more about this, this should work:
- If `window.autoDetectColorScheme` is true, have `nativeTheme.themeSource = 'system'`. The window appearance as well as the workbench color theme will be in sync as the workbench color theme follows the OS appearance.
- If `window.autoDetectColorScheme` is false, observe the current color theme and set  `nativeTheme.themeSource = 'dark'` or  `nativeTheme.themeSource = 'light'` so the window appearance matches the color theme

This could be done in the themeMainService.
If needed there can also be an extra setting to control whether changing the window appearance based on color theme is wanted.

What needs investigation is how this affects the detection of high contrast

**Comment by bpasero** [comment]
> If window.autoDetectColorScheme is false, observe the current color theme and set nativeTheme.themeSource = 'dark' or nativeTheme.themeSource = 'light' so the window appearance matches the color theme

I am worried that this is a somewhat breaking change for everyone that just runs with default settings and might be confusing for users. If macOS is running in light mode (default), it is probably weird that VSCode suddenly appears dark, both in the window title as well as context menus. 

So I think either a new setting or an update to the existing setting is still needed.

**Comment by No-Cellist-7780** [comment]
Hi, I want to work on this issue, any guidance on how to proceed on this?

**Comment by bpasero** [comment]
With the new setting `window.systemColorTheme` you can now explicitly opt into `dark` mode even when the OS is set to `light`.

You can give our preview releases a try from: https://code.visualstudio.com/insiders/
